src/main/kotlin/com/exactpro/th2/ws/client/api/impl/WebSocketClient.kt

```diff
@@ -135,62 +136,84 @@ class WebSocketClient(
     override fun onClose(socket: WebSocket, statusCode: Int, reason: String): CompletionStage<*>? = try {
         onInfo { "Disconnected from: $uri - statusCode: $statusCode, reason: $reason" }
         handler.onClose(statusCode, reason)
-        this.socket = connect()
+        connect()
         null
     } catch (e: Exception) {
         onError(e) { "Failed to handle onClose event" }
-        this.socket = connect()
+        connect()
         null
     }
 
     override fun onError(socket: WebSocket, error: Throwable) = try {
         onError(error) { "Disconnected from: $uri" }
         handler.onError(error)
-        this.socket = connect()
+        connect()
     } catch (e: Exception) {
         onError(e) { "Failed to handle onError event" }
-        this.socket = connect()
+        connect()
     }
 
-    override fun close() {
-        logger.info { "Closing" }
-        running = false
+    @Synchronized
+    fun start() = when {
+        isRunning -> onInfo { "Client is already running" }
+        else -> {
+            onInfo { "Starting client" }
+            isRunning = true
+            connect()
+            onInfo { "Started client" }
+        }
+    }
+
+    @Synchronized
+    fun stop() = when {
+        !isRunning || !::socket.isInitialized -> onInfo { "Client is already stopped" }
+        else -> {
+            onInfo { "Stopping client" }
+
+            socket.runCatching {
+                isRunning = false
 
-        socket.runCatching {
-            if (isOutputClosed) {
-                logger.warn { "Trying to close socket abruptly" }
-                abort()
-            } else {
-                logger.info { "Trying to close socket gracefully" }
-                sendClose(WebSocket.NORMAL_CLOSURE, "")
+                if (isOutputClosed) {
+                    logger.warn { "Trying to close socket abruptly" }
+                    abort()
+                } else {
+                    logger.info { "Trying to close socket gracefully" }
+
+                    handler.runCatching(IHandler::preClose).onFailure {
+                        onError(it) { "Failed to handle preClose event" }
+                    }
+
+                    sendClose(WebSocket.NORMAL_CLOSURE, "")
+                }
+            }.onFailure {
+                logger.error(it) { "Failed to close socket" }
             }
-        }.onFailure {
-            logger.error(it) { "Failed to close socket" }
+
+            onInfo { "Stopped client" }
         }
+    }
 
-        logger.info { "Closed" }
+    private fun connect() {
+        if (!isRunning) return
+        this.socket = createSocket()
```

Comment:
the socket is not volatile, each method that works with it should be synchronized